core/appmodule/environment.go

```diff
@@ -0,0 +1,19 @@
+package appmodule
+
+import (
+	"cosmossdk.io/core/branch"
+	"cosmossdk.io/core/event"
+	"cosmossdk.io/core/gas"
+	"cosmossdk.io/core/header"
+	"cosmossdk.io/core/store"
+)
+
+// Environment is used to get all services to their respective module
+type Environment struct {
+	BranchService   branch.Service
+	EventService    event.Service
+	GasService      gas.Service
+	HeaderService   header.Service
+	KvStoreService  map[string]store.KVStoreService
```

Comment:
question i have is should this be for all modules or only the module it is passed to?

Reply:
In NewEnvironment i could make it a map of envs for each module so not every module has the other ones store key

Reply:
My gut tells me that each instance of an `Environment` passed to a module should be scoped to that module, i.e. each module can only get access to the store for its own store key. This, in a way, sort of gives us OCAP.

Reply:
Also events and branching are module scoped

Reply:
true, ill make the changes